Remove full rebuilds for tag indexer with every keystroke or UI click

pmario opened the pull request:
This PR will fix #7455

As issue #7455 shows, the tag indexes get destroyed with every keystroke in edit mode and with every user UI interaction. IMO that should not be the case. 

This PR also contains the performance measurement logic, so Vercel builds a test-version. Once the preview-version is built I'll update the PR and remove that code. 

Since the updateDescriptor object is nested and ES5 needs a lot of boilerplate code to avoid "undefined" exceptions I did create several helper variables, which should make the code much more understandable. 

As the screenshots at the issue shows, the CSS "rebuilding" speed can be doubled.
I did some test with 2000 tags and the relative performance improvement is the same.

pmario committed:
WIP first-try reduce tag-indexer overhead

pmario committed:
a bit more inline docs

pmario commented:
The following text is copied over from #7455 

--------

I did some initial test with this simple logic, which tests if the `tags` field exists. No further processing. tiddlywiki.com has 388 tags atm. ... So I would need to test it with more tags 2000 or more. 

<details><summary>show code <--- </summary>

```
TagIndexer.prototype.update = function(updateDescriptor) {
	if( (!updateDescriptor.new.tiddler?.fields?.tags && updateDescriptor.old.tiddler === undefined) ||
		(!updateDescriptor.old.tiddler?.fields?.tags && updateDescriptor.new.tiddler === undefined) ||
		(!updateDescriptor.old.tiddler?.fields?.tags && !updateDescriptor.new.tiddler?.fields?.tags)) {
			return;
		} else {
			$tw.utils.each(this.subIndexers,function(subIndexer) {
				subIndexer.update(updateDescriptor);
			});
		}
};
```
</details>

Then I did measure, how long a rebuild takes. 

<details><summary>show code <--- </summary>

```
TagSubIndexer.prototype.rebuild = function() {
	var self = this;
	// Hashmap by tag of array of {isSorted:, titles:[]}
const t0 = $tw.utils.timer();
	this.index = Object.create(null);
	// Add all the tags
	this.indexer.wiki[this.iteratorMethod](function(tiddler,title) {
		$tw.utils.each(tiddler.fields.tags,function(tag) {
			if(!self.index[tag]) {
				self.index[tag] = {isSorted: false, titles: [title]};
			} else {
				self.index[tag].titles.push(title);
			}
		});
	});
console.log("--------------- rebuild", $tw.utils.timer(t0));
};
```
</details>

The TW startup shows this. Every rebuild on my machine needs between 1ms and 3ms (worst case)

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/ba6965b2-7f79-45cb-a5e0-cb2b99610472)

That's not as much as I thought, but every "styleRefresh" needs about 2.5ms now. **That is about 2ms better with basically every click on the UI.** 

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/2fafbf69-1641-4fdc-a7c6-f4c0594a5472)

v5.2.7 code needs about double the time. 

![image](https://github.com/Jermolene/TiddlyWiki5/assets/374655/7996e8c7-1ea4-453f-a576-aa7e4c332c12)

-----

My machine is an  **Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz   3.60 GHz** ... Not the newest machine but probably 10 times faster than an average mobile device.

pmario committed:
tag-indexer - remove perf timing code

linonetwo commented:
Great, in most of the times, `there is no new tiddler` will be the case.

pmario commented on core/modules/indexers/tag-indexer.js:
```diff
@@ -35,9 +35,27 @@ TagIndexer.prototype.rebuild = function() {
+	var newTid = updateDescriptor.new.tiddler;
+	var oldTid = updateDescriptor.old.tiddler;
+	var noOldTid = oldTid === undefined;
+	var noNewTid = newTid === undefined;
```
We cannot return here. The if-clause is needed to know if we should reset the indexers. The vars a, b and c are there for increased readability. I don not want to pre-mature optimize them away. It does not matter.

linonetwo replied:
But you have noNewTid in the if below, and is in "or" statement

linonetwo replied:
> The vars a, b and c are there for increased readability

Put them in separate if statement with comment does the same

pmario commented:
>Great, in most of the times, there is no new tiddler will be the case.

I see what you mean. I'll ad an extra `if (noNewTid) return` after the first variable definitions to avoid the a,b,c evaluation

pmario committed:
simplify "keep subIndexer state" logic

pmario commented:
@linonetwo .. I did implement the "return early" logic and it works for me. Can you test and also make some measurements. If it makes a difference for your projects?

The code I used for measurements is documented at: https://github.com/Jermolene/TiddlyWiki5/pull/7615#issuecomment-1642078800 behind the "show code buttons"

linonetwo commented on core/modules/indexers/tag-indexer.js:
```diff
@@ -35,9 +35,27 @@ TagIndexer.prototype.rebuild = function() {
+	} //  -> or
+	// a) new tiddler has no tags && no old tiddler -> or
+	// b) new tiddler has no tags && old tiddler has no tags
+	var a=((newTid && newTid.fields && newTid.fields.tags === undefined) && noOldTid),
```
You can return early whenever possible. Here just 

```js
// a) new tiddler has no tags && no old tiddler -> or
if ((newTid && newTid.fields && newTid.fields.tags === undefined) && noOldTid) return;
// b) new tiddler has no tags && old tiddler has no tags
if ((newTid && newTid.fields && newTid.fields.tags === undefined) && (oldTid && oldTid.fields && oldTid.fields.tags === undefined) return;

pmario replied:
These micro optimisations do not make sense. I did add this code to improve readability and make it easier to understand the logic.

linonetwo commented on core/modules/indexers/tag-indexer.js:
```diff
@@ -35,9 +35,27 @@ TagIndexer.prototype.rebuild = function() {
+
+	if( a || b ) {
+			return;
+		} else {
```
https://eslint.org/docs/latest/rules/no-else-return

I thought we already have setup eslint? Did you get the warning?

pmario replied:
I do not get any warning. The eslint configuration we discussed for pretty printing is not active yet.

linonetwo commented:
I don't know where I can test it better than you, I think code is good, I can put the prerelease into TidGi after this is merged.

pmario committed:
simplify if - return - else code

Jermolene commented:
Hi @pmario looking at this again, I think the logic is faulty. The function bails early if it receives a deletion, which seems wrong. Let's look at this again after v5.3.2

pmario committed:
Merge branch 'master' into tag-indexer-perf-7455

pmario committed:
update caches if tiddler with tag has been deleted

pmario commented:
converted this PR into a draft, since it contains `console.log()` code, which should not be part of the PR

pmario commented:
The ZIP file contains 2 single-file wikis. 

- `tag-index.html` contains the "old code" + console.log() - which updated the tag-caches with every keystroke.
- `tag-optimized.html` contains the "new code" + console.log() - but it avoids tag-cache updates for most UI interactions. 

- both wikis contain 20000 new tiddlers with 20000 new tags
  - **WARNING** If you open the Right Sidebar -> More -> Tags tab it may brick your browser. 
  - It needs about 12 seconds on my machine. 

[tag-indexer-performance-measurements.zip](https://github.com/Jermolene/TiddlyWiki5/files/13476638/tag-indexer-performance-measurements.zip)

The best way to see the differences is, to open the browser dev-tools with F12 and interact with the UI. 

- tag-index.html will activate a lot more cache-updates
- styleRefresh will need about 90% more time with every interaction
- At browser reload tag-index.html on my machine needs about 80ms more time because of unnecessary cache refreshes

pmario committed:
Merge branch 'master' into tag-indexer-perf-7455

pmario committed:
Merge branch 'master' into tag-indexer-perf-7455

pmario committed:
make eslint happy

pmario committed:
make eslint and me happy

pmario committed:
also check for list field if there are no tags int now and old

pmario committed:
Merge branch 'master' into tag-indexer-perf-7455

pmario committed:
Tag indexer: skip index rebuild when tags and list are unchanged

- TagIndexer.update now returns early when neither the `tags` nor the `list` field changed, keeping such lookups O(1) instead of O(store).
- At 40,000 tiddlers this removes about 20ms of work per keystroke on a tagged tiddler.
- Add test-tag-indexer.js, a regression test asserting the index is not rebuilt on an unrelated edit and is still rebuilt when tags, list, or tiddler existence change.
- Add an archived performance benchmark under editions/test/tiddlers/tests/performance/ that measures the rebuild cost and auto-disables in CI/CD and from core v5.5.0.
